source/common/event/dispatcher_impl.cc

```diff
@@ -162,14 +162,14 @@ void DispatcherImpl::run(RunType type) {
 }
 
 void DispatcherImpl::runPostCallbacks() {
-  std::unique_lock<std::mutex> lock(post_lock_);
+  Thread::LockGuard lock(post_lock_);
   while (!post_callbacks_.empty()) {
     std::function<void()> callback = post_callbacks_.front();
     post_callbacks_.pop_front();
 
-    lock.unlock();
+    post_lock_.unlock();
```

Comment:
FWIW, though this is correct because callback() won't throw, this seems a little shady to me. Shouldn't all modifications to the lock technically be through the guard so that it is exception safe? Do we have a guard that tracks unlocks and relocks? I think we should use that.

Reply:
The annotation analysis has a hard time punching through guards.  What makes it work with LockGuard is the manual annotation at its interface as a SCOPED_LOCKABLE.

In this particular case, callback() can throw and it's fine; we're not locked during the callback, so I think all the same guarantees are made.

Reply:
If you throw, you will unlock() twice via the guard, right?

Reply:
Good catch. I think I'm going to revert this one too...I tried the most obvious ways of getting the annotation to work with this pattern and it doesn't.
```
class SCOPED_LOCKABLE LG {
 public:
  LG(Thread::BasicLockable& mutex) EXCLUSIVE_LOCK_FUNCTION(mutex) : lock_(mutex) { lock(); }
  ~LG() UNLOCK_FUNCTION() { unlock(); }
  void unlock() UNLOCK_FUNCTION() { lock_.unlock(); }
  void lock() EXCLUSIVE_LOCK_FUNCTION() { lock_.lock(); }
private:
  Thread::BasicLockable& lock_;
};
```
gives me:
```
ERROR: /usr/local/google/home/jmarantz/git3/envoy/source/common/event/BUILD:11:1: C++ compilation of rule '//source/common/event:dispatcher_lib' failed (Exit 1)
source/common/event/dispatcher_impl.cc:182:10: error: acquiring mutex 'lock' that is already held [-Werror,-Wthread-safety-analysis]
    lock.lock();
         ^
1 error generated.
```
at the line with `lock.lock();` after `callback();`  -- super frustrating.  let me think about how to work around this.

Reply:
OK. I will be honest in saying that I like the ideas of the annotations, but how much work/churn is required to get them working does kind of scare me. :)

Reply:
I hear you.  It'd be nice to use them strictly, but for now WDYT of just using them when straightforward?

Reply:
+1

Reply:
I took another crack at this, not by making a more complicated lock-guard but by rearranging the loop to make it easier to apply a simple lock guard.  WDYT?